projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/counters/dimensions.cpp

```diff
@@ -98,51 +91,65 @@ getBlockDimensions(std::string_view agent, const Metric& metric)
 namespace
 {
 metric_dims
-generate_dimensions()
+generate_dimensions(rocprofiler_agent_id_t agent_id)
 {
     std::unordered_map<uint64_t, std::vector<MetricDimension>> dims;
 
-    const auto asts = counters::get_ast_map();
-    for(const auto& [gfx, metrics] : asts->arch_to_counter_asts)
+    // Get the agent to determine which architecture's metrics to load
+    const auto* agent = rocprofiler::agent::get_agent(agent_id);
+    if(!agent) return {.id_to_dim = dims};
+
+    const auto  asts = counters::get_ast_map();
+    const auto* arch_asts =
+        rocprofiler::common::get_val(asts->arch_to_counter_asts, std::string(agent->name));
+    if(!arch_asts) return {.id_to_dim = dims};
+
+    for(const auto& [metric, ast] : *arch_asts)
     {
-        for(const auto& [metric, ast] : metrics)
+        auto ast_copy = ast;
+        try
         {
-            auto ast_copy = ast;
-            try
-            {
-                dims.emplace(ast.out_id().handle, ast_copy.set_dimensions());
-            } catch(std::runtime_error& e)
-            {
-                ROCP_FATAL << metric << " has improper dimensions"
-                           << " " << e.what();
-            }
+            // Generate dimensions for this specific agent
+            dims.emplace(ast.out_id().handle, ast_copy.set_dimensions(agent_id));
+        } catch(std::runtime_error& e)
+        {
+            ROCP_FATAL << metric << " has improper dimensions"
+                       << " " << e.what();
         }
     }
     return {.id_to_dim = dims};
 }
 }  // namespace
 
 std::shared_ptr<const metric_dims>
-get_dimension_cache(bool reload)
+get_dimension_cache(rocprofiler_agent_id_t agent_id, bool reload)
 {
-    using DimSync             = common::Synchronized<std::shared_ptr<const metric_dims>>;
-    static DimSync*& dim_data = common::static_object<DimSync>::construct(
-        [&]() { return std::make_shared<const metric_dims>(generate_dimensions()); }());
+    using DimSync =
+        common::Synchronized<std::unordered_map<uint64_t, std::shared_ptr<const metric_dims>>>;
```

Comment:
Please key the maps to the type, not the handle. If you include <rocprofiler-sdk/cxx/operators.hpp> and <rocprofiler-sdk/cxx/hash.hpp> it will work.